java-agent/src/main/java/com/signalfx/tracing/examples/javaagent/App.java

```diff
@@ -1,89 +1,120 @@
  * auto-instrumented).  Both of these operations are performed within a root
  * parent span called "fetch-and-set" that is manually created in the code
  * below.
- *
+ * <p>
  * See the example for the Jaeger Java tracer for more detailed information on
- * the use of the tracer.
+ * the use of the OpenTracing API.
  */
 public class App {
-  public static void main(String[] argv) {
-    // The Java agent sets the global tracer instance before this method is
-    // called so it is safe to grab and use here.
-    Tracer tracer = GlobalTracer.get();
+    private final OkHttpClient httpClient;
+    private final Jedis redisClient;
+    private final ExecutorService executor;
 
-    OkHttpClient client = new OkHttpClient.Builder().build();
+    App() {
+        this.redisClient = new Jedis("localhost");
+        this.httpClient = new OkHttpClient.Builder().build();
+        executor = Executors.newSingleThreadExecutor();
+    }
+
+    public static void main(String[] argv) {
+        if (argv.length == 0) {
+            System.out.println("Please specify a URL to fetch");
+            System.exit(1);
+        }
+
+        String url = argv[0];
 
-    if (argv.length == 0) {
-        System.out.println("Please specify a URL to fetch");
-        System.exit(1);
+        App app = new App();
+        app.doFetchAndSet(url);
     }
 
-    String url = argv[0];
+    private void doFetchAndSet(String url) {
+        Request request = new Request.Builder()
+                .url(url)
+                .build();
 
-    Request request = new Request.Builder()
-            .url(url)
-            .build();
+        // This span acts as a root span that envelops the sets of spans generated
+        // by both the HTTP call (using the OKHttp auto-instrumentation) and the
+        // Redis set action (using the redis-jedis auto-instrumentation).  If you
+        // did not have this root span that is manaully created, you would instead
+        // see two independent traces for the HTTP get and the Redis set.
+        // The Java agent sets the GlobalTracer instance before the application's main method is called.
+        Span span = GlobalTracer.get().buildSpan("fetch-and-set").start();
+        try (Scope sc = GlobalTracer.get().scopeManager().activate(span, false)) {
+            String respBody;
+            try {
+                Response response = httpClient.newCall(request).execute();
+                respBody = response.body().string();
+            } catch (IOException e) {
+                System.out.println("Error: " + e);
+                return;
+            }
 
-    Jedis redisClient = new Jedis("localhost");
+            setInRedisWithThread(url, respBody);
+        } catch (Throwable e) {
+            Tags.ERROR.set(span, true);
+            span.log(Collections.singletonMap(ERROR_OBJECT, e));
+            System.out.println(e);
+        } finally {
+            span.finish();
```

Comment:
Is there a risk for this root span finishing before the traced redis SET's does?

Reply:
Yes, but I don't know of anything in OT that requires spans to be temporally nested, it only defines child spans as causally related.  Their [span timing diagram](https://github.com/opentracing/specification/blob/master/specification.md#references-between-spans) even shows child spans extending past the time of parent spans.  Our backend doesn't have any issues with it either from what I have seen.  Do you know of anything bad that could happen in such a scenario?

Reply:
Nope that's perfectly fine, that's how async processes work:

```
[ op-that-schedules-an-async-task ]
  +------[ a-much-longer-async-task-that-may-finish-much-later-than-its-parent ]
```